Changes from svn

http://bugs.gentoo.org/72220

--- extra/scripts/getent
+++ extra/scripts/getent
@@ -16,12 +16,16 @@
 
 #aliases|ethers|group|hosts|netgroup|networks|passwd|protocols|rpc|services|shadow)
 # dns based search is not supported for hosts|networks
-# networks searches ^string
-# protocols|rpc|services searches string anywhere
-# ethers|netgroup ?
+# ethers|netgroup (not done, needed)?
 # it returns only the first match
 case $1 in
-  group|passwd|shadow)
+  passwd)
+    string="(^\<$2\>:|^.*:.*:\<$2\>:.*:.*:.*:.*)"
+    ;;
+  group)
+    string="(^|:)\<$2\>:"
+    ;;
+  shadow)
     string="^\<$2\>:"
     ;;
   aliases)
@@ -35,7 +39,7 @@
   networks)
     string="^\<$2\>"
     ;;
-  hosts|protocol|rpc|services)
+  hosts|protocols|rpc|services)
     string="\<$2\>"
     ;;
   *)
