--- uClibc-0.9.29/libm/Makefile.in~	2007-04-12 05:41:16.000000000 +0200
+++ uClibc-0.9.29/libm/Makefile.in	2008-09-14 14:43:19.000000000 +0200
@@ -62,7 +62,7 @@
 libm_CSRC := \
 	e_acos.c e_acosh.c e_asin.c e_atan2.c e_atanh.c e_cosh.c \
 	e_exp.c e_fmod.c e_gamma.c e_gamma_r.c e_hypot.c e_j0.c \
-	e_j1.c e_jn.c e_lgamma.c e_lgamma_r.c e_log.c e_log10.c \
+	e_j1.c e_jn.c e_lgamma.c e_lgamma_r.c e_log.c e_log2.c e_log10.c \
 	e_pow.c e_remainder.c e_rem_pio2.c e_scalb.c e_sinh.c \
 	e_sqrt.c k_cos.c k_rem_pio2.c k_sin.c k_standard.c k_tan.c \
 	s_asinh.c s_atan.c s_cbrt.c s_ceil.c s_copysign.c s_cos.c \
@@ -70,10 +70,11 @@
 	s_ilogb.c s_ldexp.c s_lib_version.c s_lrint.c s_lround.c s_llround.c \
 	s_log1p.c s_logb.c s_matherr.c s_modf.c s_nextafter.c s_round.c \
 	s_rint.c s_scalbn.c s_signgam.c s_significand.c s_sin.c s_tan.c \
+	s_fdim.c s_fma.c s_fmax.c s_fmin.c s_nearbyint.c s_remquo.c s_scalbln.c \
 	s_tanh.c s_trunc.c w_acos.c w_acosh.c w_asin.c w_atan2.c w_atanh.c \
-	w_cabs.c w_cosh.c w_drem.c w_exp.c w_fmod.c w_gamma.c w_gamma_r.c \
+	w_cabs.c w_cosh.c w_drem.c w_exp.c w_exp2.c w_fmod.c w_gamma.c w_gamma_r.c w_tgamma.c \
 	w_hypot.c w_j0.c w_j1.c w_jn.c w_lgamma.c w_lgamma_r.c \
-	w_log.c w_log10.c w_pow.c w_remainder.c w_scalb.c w_sinh.c \
+	w_log.c w_log2.c w_log10.c w_nexttoward.c w_pow.c w_remainder.c w_scalb.c w_sinh.c \
 	w_sqrt.c fpmacros.c nan.c carg.c s_llrint.c
 FL_MOBJ := \
 	acosf.o acoshf.o asinf.o asinhf.o atan2f.o atanf.o atanhf.o cbrtf.o \
@@ -83,7 +84,7 @@
 	log1pf.o log2f.o logbf.o logf.o lrintf.o lroundf.o modff.o nearbyintf.o \
 	nextafterf.o powf.o remainderf.o remquof.o rintf.o roundf.o \
 	scalblnf.o scalbnf.o sinf.o sinhf.o sqrtf.o tanf.o tanhf.o \
-	tgammaf.o truncf.o cargf.o llrintf.o
+	tgammaf.o truncf.o cargf.o llrintf.o gammaf.o nexttowardf.o
 else
 # This list of math functions was taken from POSIX/IEEE 1003.1b-1993
 libm_CSRC := \
--- uClibc-0.9.29/libm/float_wrappers.c~	2007-04-12 05:41:16.000000000 +0200
+++ uClibc-0.9.29/libm/float_wrappers.c	2008-09-14 15:03:54.000000000 +0200
@@ -13,20 +13,6 @@
 #include <math.h>
 #include <complex.h>
 
-/* For the time being, do _NOT_ implement these functions
- * that are defined by SuSv3 */
-#undef L_exp2f         /*float       exp2f(float);*/
-#undef L_fdimf         /*float       fdimf(float, float);*/
-#undef L_fmaf          /*float       fmaf(float, float, float);*/
-#undef L_fmaxf         /*float       fmaxf(float, float);*/
-#undef L_fminf         /*float       fminf(float, float);*/
-#undef L_log2f         /*float       log2f(float);*/
-#undef L_nearbyintf    /*float       nearbyintf(float);*/
-#undef L_nexttowardf   /*float       nexttowardf(float, long double);*/
-#undef L_remquof       /*float       remquof(float, float, int *);*/
-#undef L_scalblnf      /*float       scalblnf(float, long);*/
-#undef L_tgammaf       /*float       tgammaf(float);*/
-
 /* Implement the following, as defined by SuSv3 */
 #if 0
 float       acosf(float);
@@ -43,9 +29,14 @@
 float       coshf(float);
 float       erfcf(float);
 float       erff(float);
+float       exp2f(float);
 float       expf(float);
 float       expm1f(float);
 float       fabsf(float);
+float       fdimf(float, float);
+float       fmaf(float, float, float);
+float       fmaxf(float, float);
+float       fminf(float, float);
 float       floorf(float);
 float       fmodf(float, float);
 float       frexpf(float value, int *);
@@ -53,24 +44,33 @@
 int         ilogbf(float);
 float       ldexpf(float, int);
 float       lgammaf(float);
+long long   llrintf(float);
 long long   llroundf(float);
 float       log10f(float);
 float       log1pf(float);
+float       log2f(float);
 float       logbf(float);
 float       logf(float);
+long        lrintf(float);
 long        lroundf(float);
 float       modff(float, float *);
+float       nearbyintf(float);
 float       nextafterf(float, float);
+float       nexttowardf(float, long double);
 float       powf(float, float);
 float       remainderf(float, float);
+float       remquof(float, float, int *);
 float       rintf(float);
 float       roundf(float);
 float       scalbnf(float, int);
+float       scalblnf(float, long);
 float       sinf(float);
 float       sinhf(float);
 float       sqrtf(float);
 float       tanf(float);
 float       tanhf(float);
+float       tgammaf(float);
+float       truncf(float);
 #endif
 
 
--- uClibc-0.9.29/libm/math_private.h~	2006-03-22 21:00:25.000000000 +0100
+++ uClibc-0.9.29/libm/math_private.h	2008-09-14 15:08:00.000000000 +0200
@@ -157,6 +157,7 @@
 extern double __ieee754_acos (double) attribute_hidden;
 extern double __ieee754_acosh (double) attribute_hidden;
 extern double __ieee754_log (double) attribute_hidden;
+extern double __ieee754_log2 (double) attribute_hidden;
 extern double __ieee754_atanh (double) attribute_hidden;
 extern double __ieee754_asin (double) attribute_hidden;
 extern double __ieee754_atan2 (double,double) attribute_hidden;
