Changes taken from upstream svn

--- extra/scripts/getent
+++ extra/scripts/getent
@@ -11,9 +17,19 @@
   fi
 }
 
+if [ -z "$1" ] ; then
+  echo "getent: wrong number of arguments" 1>&2
+  exit 1
+fi
+
 file="/etc/$1"
 string="dummy"
 
+if [ ! -f "$file" ] ; then
+  echo "Unknown database: $1" 1>&2
+  exit 1
+fi
+
 #aliases|ethers|group|hosts|netgroup|networks|passwd|protocols|rpc|services|shadow)
 # dns based search is not supported for hosts|networks
 # ethers|netgroup (not done, needed)?
@@ -48,6 +64,10 @@
     ;;
 esac
 
-search_entry $file $2
+if [ -z "$2" ] ; then
+  exec cat $file
+else
+  search_entry "$file" "$2"
+fi
 
 exit $retval
