--- uClibc-0.9.27/utils/Makefile.crt~	2005-08-12 16:28:15 +0200
+++ uClibc-0.9.27/utils/Makefile	2005-08-12 16:29:31 +0200
@@ -37,6 +37,40 @@
 all: $(TARGET_ICONV)
 endif
 
+N_CRT_FILE=crt1.o
+N_START_FILE=crtbegin.o
+N_END_FILE=crtend.o
+
+ifeq ($(UCLIBC_BUILD_PIE),y)
+CRT_FILE=Scrt1.o
+START_FILE=crtbeginS.o
+END_FILE=crtendS.o
+else
+CRT_FILE=$(N_CRT_FILE)
+START_FILE=$(N_START_FILE)
+END_FILE=$(N_END_FILE)
+endif
+
+START_FILES=$(TOPDIR)lib/$(CRT_FILE) $(TOPDIR)lib/crti.o $(LIBGCC_DIR)/$(START_FILE)
+END_FILES=-L $(TOPDIR)lib -lc $(LIBGCC) $(LIBGCC_DIR)$(END_FILE) $(TOPDIR)lib/crtn.o
+
+# arm and ia64 do not have crtbeginT.o for static linking
+ifeq ($(TARGET_arm),y)
+STATIC_START_FILE=$(N_START_FILE)
+else
+STATIC_START_FILE=crtbeginT.o
+endif
+
+ifeq ($(strip $(UCLIBC_STATIC_LDCONFIG)),y)
+LDCONFIG_CFLAGS := -static
+STATIC_START_FILES=$(TOPDIR)lib/$(N_CRT_FILE) $(TOPDIR)lib/crti.o $(LIBGCC_DIR)/$(STATIC_START_FILE)
+STATIC_END_FILES=-L $(TOPDIR)lib -lc $(LIBGCC) $(LIBGCC_DIR)/$(N_END_FILE) $(TOPDIR)lib/crtn.o
+else
+LDCONFIG_CFLAGS := $(PIEFLAG) $(LDPIEFLAG)
+STATIC_START_FILES=$(START_FILES)
+STATIC_END_FILES=$(END_FILES)
+endif
+
 headers:
 	@$(LN) -fs $(TOPDIR)include/elf.h
 
@@ -46,29 +80,24 @@
 	$(CC) $(CFLAGS) $^ -o $@
 	$(STRIPTOOL) -s -x -R .note -R .comment $@
 
-ifeq ($(strip $(UCLIBC_STATIC_LDCONFIG)),y)
-LDCONFIG_CFLAGS := -static
-else
-LDCONFIG_CFLAGS := $(PIEFLAG) $(LDPIEFLAG)
-endif
 ldconfig: ldconfig.c
 	$(CC) $(CFLAGS) $(LDCONFIG_CFLAGS) \
 		-DUCLIBC_RUNTIME_PREFIX=\"$(RUNTIME_PREFIX)\" \
 		-DUCLIBC_LDSO=$(UCLIBC_LDSO) -I. -I../ldso/include \
-		$^ -o $@
+		-nostdlib -o $@ $(STATIC_START_FILES) $^ $(STATIC_END_FILES)
 	$(STRIPTOOL) -s -x -R .note -R .comment $@
 
 ldd: ldd.c
 	$(CC) $(CFLAGS) $(PIEFLAG) $(LDPIEFLAG) \
 		-DUCLIBC_RUNTIME_PREFIX=\"$(RUNTIME_PREFIX)\" \
 		-DUCLIBC_LDSO=$(UCLIBC_LDSO) -I. -I../ldso/include \
-		$^ -o $@
+		-nostdlib -o $@ $(START_FILES) $^ $(END_FILES)
 	$(STRIPTOOL) -s -x -R .note -R .comment $@
 
 iconv: ../libc/misc/wchar/wchar.c
 	$(CC) $(CFLAGS) $(PIEFLAG) $(LDPIEFLAG) \
 		-DL_iconv_main \
-		$^ -o $@
+		-nostdlib -o $@ $(START_FILES) $^ $(END_FILES)
 	$(STRIPTOOL) -s -x -R .note -R .comment $@
 
 ifeq ($(strip $(HAVE_SHARED)),y)
